### PR TITLE
[SPARK-38687][SQL] Use error classes in the compilation errors of generators

### PR DESCRIPTION
**Pull request by panbingkun**
## What changes were proposed in this pull request?
Migrate the following errors in QueryCompilationErrors onto use error classes:

- nestedGeneratorError => UNSUPPORTED_GENERATOR.NESTED_IN_EXPRESSIONS
- moreThanOneGeneratorError => UNSUPPORTED_GENERATOR.MULTI_GENERATOR
- generatorOutsideSelectError => UNSUPPORTED_GENERATOR.OUTSIDE_SELECT
- generatorNotExpectedError => UNSUPPORTED_GENERATOR.NOT_GENERATOR

### Why are the changes needed?
Porting compilation errors of generator to new error framework, improve test coverage, and document expected error messages in tests.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
By running new test:
```
$ build/sbt "sql/testOnly *QueryCompilationErrorsSuite*"
```

### CODE REVIEW

**Comment by panbingkun** [comment]
ping @MaxGekk

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryCompilationErrorsSuite.scala [hunk 1]
```diff
@@ -583,6 +583,66 @@ class QueryCompilationErrorsSuite
...
+    "generators are not supported when it's nested in expressions") {
+    val e = intercept[AnalysisException](
+      sql("""select explode(Array(1, 2, 3)) + 1""").collect()
+    )
+
+    checkErrorClass(
+      exception = e,
+      errorClass = "UNSUPPORTED_GENERATOR",
+      errorSubClass = Some("NESTED_IN_EXPRESSIONS"),
+      msg = "The generator is not supported: nested in expressions (explode(array(1, 2, 3)) + 1)")
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryCompilationErrorsSuite.scala:597]: Please, quote (explode(array(1, 2, 3)) + 1) by default (double quotes). Maybe, it is better to add a helper method `toSQLExpr()` to QueryErrorsBase

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryCompilationErrorsSuite.scala [hunk 2]
```diff
@@ -583,6 +583,66 @@ class QueryCompilationErrorsSuite
...
+    val e = intercept[AnalysisException](
+      sql("""select explode(Array(1, 2, 3)), explode(Array(1, 2, 3))""").collect()
+    )
+
+    checkErrorClass(
+      exception = e,
+      errorClass = "UNSUPPORTED_GENERATOR",
+      errorSubClass = Some("MULTI_GENERATOR"),
+      msg = "The generator is not supported: only one generator allowed per select clause " +
+        "but found 2: explode(array(1, 2, 3)), explode(array(1, 2, 3))"
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryCompilationErrorsSuite.scala:610]: Let quote expressions:
explode(array(1, 2, 3)), explode(array(1, 2, 3)) -> "explode(array(1, 2, 3))", "explode(array(1, 2, 3))"

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryCompilationErrorsSuite.scala [hunk 3]
```diff
@@ -583,6 +583,66 @@ class QueryCompilationErrorsSuite
...
+          |SELECT explodedvalue.*
+          |FROM VALUES array(1, 2, 3) AS (value)
+          |LATERAL VIEW array_contains(value, 1) AS explodedvalue""".stripMargin).collect()
+    )
+
+    checkErrorClass(
+      exception = e,
+      errorClass = "UNSUPPORTED_GENERATOR",
+      errorSubClass = Some("NOT_GENERATOR"),
+      msg = "The generator is not supported: array_contains is expected to be a generator. " +
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryCompilationErrorsSuite.scala:641]: Please, quite array_contains as SQL Id: `array_contains`

**Review by MaxGekk — APPROVED** [review]
+1, LGTM. Merging to master.
Thank you, @panbingkun.

**Comment by MaxGekk** [comment]
@panbingkun Could you backport the changes to branch-3.3, please.

**Comment by panbingkun** [comment]
> @panbingkun Could you backport the changes to branch-3.3, please.

ok, i will do it.

**Pull request closed without merging**